keps/sig-apps/2879-ready-pods-job-status/README.md

```diff
@@ -0,0 +1,323 @@
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [x] (R) Production readiness review completed
+- [x] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The Job status has a field `active` which counts the number of Job Pods that
+are in `Running` or `Pending` phases. In this KEP, we add a field `ready` that
+counts the number of Job Pods that have a `Ready` condition, with the same
+best effort guarantees as the existing `active` field.
+
+## Motivation
+
+Job Pods can remain in the `Pending` phase for a long time in clusters with
+tight resources and when image pulls take long. Since the `Job.status.active`
+field includes `Pending` Pods, this can give a false impression of progress
+to end users or other controllers. This is more important when the pods serve
+as workers and need to communicate among themselves.
+
+A separate `Job.status.ready` field can provide more information for users
+and controllers, reducing the need to listen to Pod updates themselves.
+
+Note that other workload APIs (such as ReplicaSet and StatefulSet) have a
+similar field: `.status.readyReplicas`.
+
+### Goals
+
+- Add the field `Job.status.ready` that keeps a count of Job Pods with the
+  `Ready` condition.
+
+### Non-Goals
+
+- Provide strong guarantees for the accuracy of the count. Due to the
+  asynchronous nature of k8s, there are can be more or less Pods currently
+  ready than what the count provides.
+
+## Proposal
+
+Add the field `.status.ready` to the Job API. The job controller updates the
+field based on the number of Pods that have the `Ready` condition.
+
+### Risks and Mitigations
+
+- An increase in Job status updates. To mitigate this, the field the job
+  controller only updates if the field if any of these conditions are true:
+  - The job controller is updating other status fields.
+  - The number of ready Pods equals `Job.spec.parallelism`.
+  - The increase of ready Pods is greater than or equal to 10% of
+    `Job.spec.parallelism`.
```

Comment:
Let's say that because of cluster capacity, 100% of my jobs will never have place to run. But let's say 99% has.
With the policies above it may happen that we update ready to 90%, but we will never change that to 99%. So the system isn't eventually consistent, which I think is problematic.

I think that you need another rule, i.e. when something changes, it will be applied with X seconds/minutes (i.e. we batch it for such period).

[FWIW - such logic is super simple to implement, e.g. https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpoint/endpoints_controller.go#L224 ]

Reply:
True, being eventually consistent should be a requirement.

Could the solution for the job controller be the same? We would delay/accumulate any sync coming from Pod creation/updates/deletions. This might actually be good for the overall performance of the controller. The delay for endpoint slices is configurable. Should we do the same? I'm proposing a 100ms window otherwise.